pkg/ARCHITECTURE.md

```diff
@@ -0,0 +1,43 @@
+| /pkg/cmd | The binaries that we build. Grafana-server and grafana-cli |
+| /pkg/components | Mixed content of packages that we copied into Grafana and packages we implemented ourself. The Purpose of this folders should be packages that are too big for util and doesnt belong somewhere else |
+| /pkg/infra | Packages in infra should be packages that are used in multiple places of Grafana without knowing anything about Grafana domain. E.g. Logs, Metrics, Traces |
+| /pkg/services | Packages in services are responsible for peristing domain objects and manage the relationship between domain objects. |
+| /pkg/tsdb | All backend implementations of the datasources in Grafana. Used by both Grafanas frontend and alerting. |
+| /pkg/util | Small helper functions that are used in multiple parts of the codebase. Many functions are placed directly in the util folders which is something we want to avoid. Its better to give the util function a more descriptive package name. Ex `errutil` |
+
+## Central components of Grafanas backend
+
+| package | description | 
+| ------- | ----------- |
+| /pkg/bus | The bus! Described more below |
+| /pkg/models | This is where we keep our domain model. This package should not depend on any package outside standard library. (It does contain some refs within Grafana but that is something we should avoid going forward) |
+| /pkg/registry | service management package. |
+| /pkg/services/alerting | Grafanas alerting services. The alerting engine run in a seperate go routine and should not depend on anything else within grafan. |
+| /pkg/services/sqlstore | Currently where are database calls resides. |
+| /pkg/setting | settings package for Grafana. Anything related to grafana global configuration should be dealt with in this package. |
+
+## Testing
+We value clean & readable code that is loosely coupled and covered by unit tests. This makes it easier to collaborate and maintain the code. In the sqlstore package we do database operations in tests and while some might say that's not suited for unit tests. We think they are fast enough and provide a lot of value. 
+
+The majority of our tests uses go convey but standard library tests are fine as well. 
+
+## The Bus
+The bus is our way to introduce indirection between the http handlers and sqlstore (responsible for fetching data from the database). Http handlers and sqlstore don't depend on each other. They only depend on the bus and the domain model(pkg/models). This makes it easier to test the code and avoids coupling. More about this under `current rewrite/refactorings`
+
+## Services/Repositories 
+Services within Grafana should be self-contained and only talk to other parts of Grafana using the bus or repositories that have been made available thru Grafana service registry. All services should register themselves to the `registry` package in an init function. Only registration should be done in the init function. Init functions should be avoided as much as possible. 
```

Comment:
We have pretty good documentation of services signature in the `registry`, but I wonder if we should mentioned it here. Or even move it to its own part.

Like for me, there was some questions like so - https://github.com/grafana/grafana-enterprise/pull/101#discussion_r292326673.

Or the bit here - https://github.com/grafana/grafana-enterprise/pull/86/files
Where I was accessing the `settings` in the `Init` and at that point settings were not initialised yet :).

It was even was missed during the review. We would caught up to eventually and we did, but it did confuse me a bit